src/themis/secure_cell_seal_passphrase.c

```diff
@@ -0,0 +1,341 @@
+                                        sizeof(derived_key),
+                                        hdr.iv,
+                                        hdr.iv_length,
+                                        user_context,
+                                        user_context_length,
+                                        message,
+                                        message_length,
+                                        encrypted_message,
+                                        encrypted_message_length,
+                                        auth_tag,
+                                        &auth_tag_length);
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+    /*
+     * We should have allocated just the right amount here, but if our guess
+     * was off then the header is not usable due to bothched data offsets.
+     */
+    if (auth_tag_length != hdr.auth_tag_length) {
+        res = THEMIS_FAIL;
+        goto error;
+    }
+
+    if (*auth_token_length < themis_scell_auth_token_passphrase_size(&hdr)) {
+        res = THEMIS_BUFFER_TOO_SMALL;
+        goto error;
+    }
+    res = themis_write_scell_auth_token_passphrase(&hdr, auth_token, *auth_token_length);
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+    res = themis_write_scell_pbkdf2_context(&hdr, &kdf, auth_token, *auth_token_length);
+    if (res != THEMIS_SUCCESS) {
+        goto error;
+    }
+    *auth_token_length = themis_scell_auth_token_passphrase_size(&hdr);
+    *encrypted_message_length = message_length;
+
+error:
+    soter_wipe(iv, sizeof(iv));
+    soter_wipe(salt, sizeof(salt));
+    soter_wipe(auth_tag, sizeof(auth_tag));
+    soter_wipe(derived_key, sizeof(derived_key));
+
+    return res;
+}
+
+themis_status_t themis_auth_sym_encrypt_message_with_passphrase(const uint8_t* passphrase,
+                                                                size_t passphrase_length,
+                                                                const uint8_t* message,
+                                                                size_t message_length,
+                                                                const uint8_t* user_context,
+                                                                size_t user_context_length,
+                                                                uint8_t* auth_token,
+                                                                size_t* auth_token_length,
+                                                                uint8_t* encrypted_message,
+                                                                size_t* encrypted_message_length)
+{
+    THEMIS_CHECK_PARAM(passphrase != NULL && passphrase_length != 0);
+    THEMIS_CHECK_PARAM(message != NULL && message_length != 0);
+    if (user_context != NULL) {
+        THEMIS_CHECK_PARAM(user_context_length != 0);
+    } else {
+        THEMIS_CHECK_PARAM(user_context_length == 0);
```

Comment:
do we need to enforce this if we have no user context? or we want our callers to be explicit - that is avoid accidentally passing NULL to the context when they did not mean to?

Reply:
I'm not sure... At the moment I don't remember what my precise reasoning was when I was writing those lines. Normally, I wouldn't expect users to reuse the same context buffer with different lengths. This is one case when it is reasonable to have `user_context != NULL` but `user_context_length == 0` and expect the operation to succeed.

Another consideration is resilience. We can get the above combination if the user messes up parameter order, like omits one of them or something. However, since no other parameter is optional or accepts zeros, this situation is likely to cause a failure and will be noticed.

Yet another consideration is that some language runtimes do not return NULL data pointer for “empty” container. They return some non-NULL pointer (which you still are not allowed to dereference). I can't speak without checking for others, but C++'s `std::vector<T>::data()` and Rust's `std::[T]::as_ptr()` behave like that.

So it does make sense to allow non-NULL `user_context` when `user_context_length` is zero.